src/graphepp/graphepp.py

```diff
@@ -52,6 +52,92 @@ def adj_matrix(N, E):
     return adj
 
 
+def local_PauliZ(graph, n):
+    """Performs at qubit n a local Pauli Z measurement within the graph state.
+
+    The graph 'graph' has `N` vertices (labeled 0 to N-1) and edges `E`.
+
+    Parameters
+    ----------
+    graph : Graph
+        Instance of Graph class
+    n : int
+        index of the qubit on which the local Pauli Z measurement is performed
+
+    Returns
+    -------
+    Graph : Graph
+        Returns a new Graph state with updated edge set
+
+    """
+
+    if n < 0 or n >= graph.N:
+        raise ValueError("qubit index out of range: ", str(n))
+    edges = list(graph.E)
+    newEdges = tuple((x, y) for (x, y) in edges if x != n and y != n)
+    return Graph(graph.N, E=newEdges)
+
+
+def local_PauliY(graph, n):
+    """Performs at qubit n a local Pauli Y measurement within the graph state.
+
+    The graph 'graph' has `N` vertices (labeled 0 to N-1) and edges `E`.
+
+    Parameters
+    ----------
+    graph : Graph
+        Instance of Graph class
+    n : int
+        index of the qubit on which the local Pauli Y measurement is performed
+
+    Returns
+    -------
+    Graph : Graph
+        Returns a new Graph state with updated edge set
+
+    """
+    loc_graph = local_complementation(n, graph)
+    return local_PauliZ(loc_graph, n)
+
+
+def local_PauliX(graph, n, neighbor=-1):
+    """Performs at qubit n a local Pauli X measurement within the graph state.
+
+    The graph 'graph' has `N` vertices (labeled 0 to N-1) and edges `E`.
+
+    Parameters
+    ----------
+    graph : Graph
+        Instance of Graph class
+    n : int
+        index of the qubit on which the local Pauli Y measurement is performed
+    neighbor : int
+        neighboring qubit on which the local complementation is performed, optional parameter
+    Returns
+    -------
+    Graph : Graph
+        Returns a new Graph state with updated edge set
+
+    """
+    # find all neighbors of qubit n
+    neighbors = np.array([(x, y) for (x, y) in list(graph.E) if x == n or y == n])
+
+    # if the vertex has neighbors, we perform the measurement, else we return the same graph state
+    if neighbors.shape[0] != 0:
+        # if neighbor is not specified, choose the first neighbor in the array
+        if neighbor == -1:
+            neighbor = neighbors[0][0] if neighbors[0][0] != n else neighbors[0][1]
```

Comment:
This seems like a lot of unnecessary index shuffling. I would also add a protection against specifying an invalid neighbor.
```suggestion
    neighbors = np.nonzero(graph.adj[n, :]
    if neighbor == -1:
        neighbor = neighbors[0]
    else:
        if neighbor not in neighbors:
            raise ValueError(f"neighbor={neighbor} is not a neighbor of vertex n={n}.")
```